File: social-media/linkedin_post.js
#!/usr/bin/env node
/**
 * LinkedIn image-upload + post script for CCN Social Hub.
 * Called by n8n Execute Command node. Reads post data from env vars.
 * Uses built-in fetch (Node 18+) — no npm packages required.
 *
 * Steps: register upload slot → download image → upload → post article
 */

const fs   = require('fs');
const path = require('path');

const MEMBER_ID = '41q6-gkeGG';
const LI_API    = 'https://api.linkedin.com';

function loadEnv() {
  // Prefer process.env (set on Docker container) over .env file
  if (process.env['LINKEDIN_ACCESS_TOKEN']) return process.env;
  const env = {};
  const raw = fs.readFileSync('/root/.hermes/.env', 'utf8');
  for (const line of raw.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
    const eqIdx = trimmed.indexOf('=');
    const k = trimmed.slice(0, eqIdx).trim();
    const v = trimmed.slice(eqIdx + 1).trim();
    env[k] = v;
  }
  return env;
}

function liHeaders(token, contentType = 'application/json') {
  return {
    'Authorization': `Bearer ${token}`,
    'LinkedIn-Version': '202507',
    'X-Restli-Protocol-Version': '2.0.0',
    'Content-Type': contentType,
  };
}

async function registerImage(token) {
  const res = await fetch(`${LI_API}/rest/images?action=initializeUpload`, {
    method: 'POST',
    headers: liHeaders(token),
    body: JSON.stringify({ initializeUploadRequest: { owner: `urn:li:person:${MEMBER_ID}` } }),
  });
  if (!res.ok) throw new Error(`Register failed: ${res.status} ${await res.text()}`);
  const data = await res.json();
  return { uploadUrl: data.value.uploadUrl, imageUrn: data.value.image };
}

async function downloadImage(imageUrl) {
  const res = await fetch(imageUrl);
  if (!res.ok) throw new Error(`Download failed: ${res.status}`);
  const buf = await res.arrayBuffer();
  return Buffer.from(buf);
}

async function uploadImage(token, uploadUrl, imageBuffer) {
  const res = await fetch(uploadUrl, {
    method: 'PUT',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'image/jpeg',
    },
    body: imageBuffer,
  });
  if (!res.ok) throw new Error(`Upload failed: ${res.status} ${await res.text()}`);
}

async function postArticle(token, { message, postUrl, title, subtitle }) {
  const mediaItem = {
    status: 'READY',
    originalUrl: postUrl,
    title: { text: title },
    description: { text: subtitle || '' },
  };

  const body = {
    author: `urn:li:person:${MEMBER_ID}`,
    lifecycleState: 'PUBLISHED',
    specificContent: {
      'com.linkedin.ugc.ShareContent': {
        shareCommentary: { text: message },
        shareMediaCategory: 'ARTICLE',
        media: [mediaItem],
      },
    },
    visibility: {
      'com.linkedin.ugc.MemberNetworkVisibility': 'PUBLIC',
    },
  };

  const bodyStr = JSON.stringify(body);
  process.stderr.write(JSON.stringify({
    debug: 'postArticle body',
    commentary: message,
    commentaryLength: message.length,
    bodyLength: bodyStr.length,
  }) + '\n');

  const res = await fetch(`${LI_API}/v2/ugcPosts`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'X-Restli-Protocol-Version': '2.0.0',
      'Content-Type': 'application/json',
    },
    body: bodyStr,
  });
  if (!res.ok) throw new Error(`Post failed: ${res.status} ${await res.text()}`);
  const data = await res.json();
  return data.id || '';
}

async function main() {
  const env   = loadEnv();
  const token = env['LINKEDIN_ACCESS_TOKEN'];

  let title, subtitle, postUrl, message;

  if (process.env['CCN_PAYLOAD']) {
    // JSON payload from n8n Execute Command — handles newlines + special chars safely
    const p = JSON.parse(Buffer.from(process.env['CCN_PAYLOAD'], 'base64').toString('utf8'));
    title    = p.title    || 'New from CCN';
    subtitle = p.subtitle || '';
    postUrl  = p.url      || 'https://danova.substack.com';
    // Use explicit message field; never fall back silently (would drop hashtags)
    if (!p.message) throw new Error('CCN_PAYLOAD missing message field');
    message  = p.message;
  } else {
    title    = process.env['LI_TITLE']     || 'New from CCN';
    subtitle = process.env['LI_SUBTITLE']  || '';
    postUrl  = process.env['LI_URL']       || 'https://danova.substack.com';
    message  = process.env['LI_MESSAGE']   || title;
  }

  process.stderr.write(JSON.stringify({
    debug: 'linkedin_post.js received',
    title,
    messagePreview: message.slice(0, 80) + (message.length > 80 ? '…' : ''),
    messageLength: message.length,
    hasHashtags: message.includes('#'),
  }) + '\n');

  const postId = await postArticle(token, { message, postUrl, title, subtitle });
  process.stdout.write(JSON.stringify({ linkedInPostId: postId, success: true }) + '\n');
}

main().catch(err => {
  process.stderr.write(JSON.stringify({ error: err.message }) + '\n');
  process.exit(1);
});

File: social-media/linkedin_post.py
#!/usr/bin/env python3
"""
LinkedIn image-upload + post script for CCN Social Hub.
Called by n8n Execute Command node. Reads post data from env vars.
Handles: register upload slot → download image → upload to LinkedIn → post article.
"""
import json, os, sys, requests

MEMBER_ID = "41q6-gkeGG"
LI_API    = "https://api.linkedin.com"

def load_env():
    env = {}
    with open("/root/.hermes/.env") as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                k, _, v = line.partition("=")
                env[k.strip()] = v.strip()
    return env

def li_headers(token, binary=False):
    h = {
        "Authorization": f"Bearer {token}",
        "LinkedIn-Version": "202507",
        "X-Restli-Protocol-Version": "2.0.0",
    }
    if not binary:
        h["Content-Type"] = "application/json"
    return h

def register_image(token):
    r = requests.post(
        f"{LI_API}/rest/images?action=initializeUpload",
        headers=li_headers(token),
        json={"initializeUploadRequest": {"owner": f"urn:li:person:{MEMBER_ID}"}},
        timeout=30,
    )
    r.raise_for_status()
    data = r.json()["value"]
    return data["uploadUrl"], data["image"]

def download_image(image_url):
    r = requests.get(image_url, timeout=30)
    r.raise_for_status()
    return r.content

def upload_image(token, upload_url, image_bytes):
    h = li_headers(token, binary=True)
    h["Content-Type"] = "image/jpeg"
    r = requests.put(upload_url, headers=h, data=image_bytes, timeout=60)
    r.raise_for_status()

def post_article(token, message, post_url, title, subtitle, image_urn=None):
    body = {
        "author": f"urn:li:person:{MEMBER_ID}",
        "commentary": message,
        "visibility": "PUBLIC",
        "distribution": {"feedDistribution": "MAIN_FEED"},
        "lifecycleState": "PUBLISHED",
        "isReshareDisabledByAuthor": False,
    }
    if image_urn:
        body["content"] = {
            "article": {
                "source": post_url,
                "thumbnail": image_urn,
                "title": title,
                "description": subtitle or "",
            }
        }
    r = requests.post(
        f"{LI_API}/rest/posts",
        headers=li_headers(token),
        json=body,
        timeout=30,
    )
    r.raise_for_status()
    return r.headers.get("x-restli-id", "")

def main():
    env       = load_env()
    token     = env["LINKEDIN_ACCESS_TOKEN"]
    title     = os.environ.get("LI_TITLE", "New from CCN")
    subtitle  = os.environ.get("LI_SUBTITLE", "")
    post_url  = os.environ.get("LI_URL", "https://danova.substack.com")
    image_url = os.environ.get("LI_IMAGE_URL", "")
    message   = os.environ.get("LI_MESSAGE", title)

    image_urn = None

    if image_url:
        try:
            upload_url, image_urn = register_image(token)
            image_bytes = download_image(image_url)
            upload_image(token, upload_url, image_bytes)
        except Exception as e:
            print(json.dumps({"warning": f"Image upload failed: {e}, posting without image"}), file=sys.stderr)
            image_urn = None

    post_id = post_article(token, message, post_url, title, subtitle, image_urn)
    print(json.dumps({"linkedInPostId": post_id, "imageUrn": image_urn, "success": True}))

if __name__ == "__main__":
    main()
